compiler/hash-typecheck/src/new/ops/bootstrap.rs

```diff
@@ -0,0 +1,205 @@
+//! This module contains code that can bootstrap the typechecker, by creating
+//! and injecting primitive definitions into the context.
+use std::{cell::Cell, iter::once};
+
+use derive_more::Constructor;
+use hash_types::new::{
+    data::{DataDefId, ListCtorInfo, NumericCtorInfo, PrimitiveCtorInfo},
+    defs::DefParamGroupData,
+    environment::{context::ScopeKind, env::AccessToEnv},
+    mods::{ModDefData, ModDefId, ModKind, ModMemberData, ModMemberValue},
+    params::ParamData,
+};
+use hash_utils::store::Store;
+
+use super::{common::CommonOps, AccessToOps};
+use crate::{
+    impl_access_to_tc_env,
+    new::environment::tc_env::{AccessToTcEnv, TcEnv},
+};
+
+#[derive(Constructor)]
+pub struct BootstrapOps<'tc> {
+    tc_env: &'tc TcEnv<'tc>,
+}
+
+macro_rules! defined_primitives {
+    ($($name:ident: $type:ty),* $(,)?) => {
+        #[derive(Debug, Copy, Clone)]
+        pub struct DefinedPrimitives {
+            $($name: DataDefId),*
+        }
+
+        impl DefinedPrimitives {
+            $(
+                pub fn $name(&self) -> DataDefId {
+                    self.$name
+                }
+            )*
+        }
+
+        impl<'tc> BootstrapOps<'tc> {
+            /// Create a list of [`ModMemberData`] that corresponds to the defined primitives.
+            ///
+            /// This can be used to make a module and enter its scope.
+            pub fn make_primitive_mod_members(&self, primitives: &DefinedPrimitives) -> Vec<ModMemberData> {
+                vec![
+                    $(
+                        ModMemberData {
+                            name: self.stores().data_def().map_fast(primitives.$name, |def| def.name),
+                            value: ModMemberValue::Data(primitives.$name)
+                        },
+                    )*
+                ]
+            }
+        }
+    };
+
+}
+
+// All the primitive types:
+defined_primitives! {
+    never: DataDefId,
+    bool: DataDefId,
```

Comment:
Will any types term with a primitive type always have this `DataDefId`, if it would then it would be very convenient to check for, and to use in later stages like IR and codegen.

Reply:
Indeed, after TC all primitives will have type Ctor(DataDefId, Args).